spyder/preferences/shortcuts.py

```diff
@@ -214,118 +292,82 @@ def accept(self):
-                self.label_warning.setText(_("Please introduce a different "
-                                             "shortcut"))
-
-            if len(self.keys) == 1 and key in [Qt.Key_Return, Qt.Key_Enter]:
-                self.toggle_state()
-                return
-
-            if not self.edit_state:
-                self.nonedit_keyrelease(e)
-            else:
-                debug_print('keys: {}'.format(self.keys))
-                if self.keys and key != Qt.Key_Escape:
-                    self.validate_sequence()
-                self.keys = set()
-                self.key_modifiers = set()
-                self.key_non_modifiers = list()
-                self.key_text = list()
-                self.npressed = 0
+        self.new_subseq.append(event_keystr)
+        self.set_sequence(', '.join(self.new_subseq))
 
     def check_conflicts(self):
         """Check shortcuts for conflicts."""
         conflicts = []
-        for index, shortcut in enumerate(self.shortcuts):
-            sequence = str(shortcut.key)
-            if sequence == self.new_sequence and \
-                (shortcut.context == self.context or shortcut.context == '_' or
-                 self.context == '_'):
-                conflicts.append(shortcut)
+        if self.new_qsequence.isEmpty():
+            return conflicts
+
+        for shortcut in self.shortcuts:
+            shortcut_qsequence = QKeySequence.fromString(str(shortcut.key))
+            if shortcut_qsequence.isEmpty():
+                continue
+            if (shortcut.context, shortcut.name) == (self.context, self.name):
+                continue
+            if shortcut.context in [self.context, '_'] or self.context == '_':
+                if (shortcut_qsequence.matches(self.new_qsequence) or
+                        self.new_qsequence.matches(shortcut_qsequence)):
+                    conflicts.append(shortcut)
         return conflicts
 
+    def check_singlekey(self):
+        """Check if the first sub-sequence of the new key sequence is valid."""
+        if self.new_qsequence.isEmpty():
+            return True
+        else:
+            keystr = self.new_subseq[0]
+            valid_single_keys = (EDITOR_SINGLE_KEYS if
+                                 self.context == 'editor' else SINGLE_KEYS)
+            if any((m in keystr for m in ('Ctrl', 'Alt', 'Shift', 'Meta'))):
+                return True
+            else:
+                # This means that the the first subsequence is composed of
+                # a single key with no modifier.
+                valid_single_keys = (EDITOR_SINGLE_KEYS if
+                                     self.context == 'editor' else SINGLE_KEYS)
+                if any((k == keystr for k in valid_single_keys)):
+                    return True
+                else:
+                    return False
+
+    def check_shift_keyseq(self):
+        """
+        Check if the first sub-sequence of the new key sequence is composed

```

Comment:
See comment regarding PEP 257.

Reply:
You mean adding a one-line summary to a two-line docstring? I don't see the added value here. I mean, I get it for a public API or a Library where the docstrings are used to actually build some public documentation, but for documenting the code source of a GUI application? I only see noise and wasted space...

If @ccordoba12 starts enforcing this in all new PR, I'll happily conform to this rule though :P

Reply:
Yeah, you have a point. I know they hasn't been really enforced in similar situations (two-line docstrings) for tests, since, well, they're only tests after all. In our contributing guide/other resources, we do specify following PEP 257, which states:

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description. The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line.

Of course, following the ethos of Python, the PEP also reminds us at the top,

> The PEP contains conventions, not laws or syntax. If you violate these conventions, the worst you'll get is some dirty looks. But some software [...] will be aware of the conventions, so following them will get you the best results.

So yeah, up to @ccordoba12 what to do. Either way, it sets a bit of a precedent to follow going forward.

Reply:
I'm with Jean on this one. Most of the docstrings we add are informative for other core developers but not meant for public consumption. So we can relax PEP 257 in our codebase (I'm not following either).

Reply:
Okay, sure, if you think so.